index.js

```diff
@@ -0,0 +1,45 @@
+import { preprocess } from './lib/preprocess.js'
+import { evaluate } from './lib/evaluate.js'
+
+export const startEvaluate = ({
+  ieContract,
+  ieContractWithSigner,
+  web3Storage,
+  logger
+}) => {
+  const rounds = {}
+  const cidsSeen = []
+  const roundsSeen = []
+
+  // Listen for events
+  ieContract.on('MeasurementsAdded', (cid, _roundIndex) => {
+    const roundIndex = Number(_roundIndex)
+    if (cidsSeen.includes(cid)) return
+    cidsSeen.push(cid)
```

Comment:
IIUC, this array will keep growing until we run out of memory.

Have you considered remembering only the last N items, where N is for example 1000?

Also, what happens when we restart the service, e.g. after deploying a new version? The list of CIDs seen will be lost. How is that going to impact the evaluation?

Reply:
Added a fix 👍 

The service restarting isn't a problem, we're mostly debouncing here anyway
- if we preprocess redundantly, there will be extra work but the results won't change
- if we evaluate redundantly, there will be extra work and gas, but the results won't change